### ISSUE TITLE
[Question]: The PMMl model was not rearranged when parsing the decision tree

### ISSUE THREAD

**Issue by luxr123** [jpmml/jpmml-evaluator#87]
```java
Map<FieldName, Object>    params  = new HashMap<>();
params.put(new FieldName("fld1"), 30.0);
params.put(new FieldName("fld2"), 60.0);
params.put(new FieldName("fld3"), "false");
params.put(new FieldName("fld4"), "optA");
```
the target fld5 result of jpmml-evaluator is "null"
but the target fld5 result of other software is "tgtY"

In fact, in the implementation process, pmml model did not do rearrangement,
here mainly fld4 logic to determine the problem  

```xml
<PMML version="4.2" xsi:schemaLocation="http://www.dmg.org/PMML-4_2 http://www.dmg.org/v4-1/pmml-4-2.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.dmg.org/PMML-4_2">
  <DataDictionary numberOfFields="5">
    <DataField dataType="double" name="fld1" optype="continuous"/>
    <DataField dataType="double" name="fld2" optype="continuous"/>
    <DataField dataType="string" name="fld3" optype="categorical">
      <Value value="true"/>
      <Value value="false"/>
    </DataField>
    <DataField dataType="string" name="fld4" optype="categorical">
      <Value value="optA"/>
      <Value value="optB"/>
      <Value value="optC"/>
    </DataField>
    <DataField dataType="string" name="fld5" optype="categorical">
      <Value value="tgtX"/>
      <Value value="tgtY"/>
      <Value value="tgtZ"/>
    </DataField>
  </DataDictionary>
  <TreeModel functionName="classification" modelName="TreeTest">
    <MiningSchema>
      <MiningField name="fld1"/>
      <MiningField name="fld2"/>
      <MiningField name="fld3"/>
      <MiningField name="fld4"/>
      <MiningField name="fld5" usageType="target"/>
    </MiningSchema>
    <Node score="tgtX">
      <True/>
      <Node score="tgtX">
        <SimplePredicate field="fld4" operator="equal" value="optA"/>
        <Node score="tgtX">
          <CompoundPredicate booleanOperator="surrogate">
            <SimplePredicate field="fld1" operator="lessThan" value="30"/>
            <SimplePredicate field="fld2" operator="greaterThan" value="20"/>
          </CompoundPredicate>
          <Node score="tgtX">
            <SimplePredicate field="fld2" operator="lessThan" value="40"/>
          </Node>
          <Node score="tgtZ">
            <SimplePredicate field="fld2" operator="greaterOrEqual" value="10"/>
          </Node>
        </Node>
        <Node score="tgtZ">
          <CompoundPredicate booleanOperator="or">
            <SimplePredicate field="fld1" operator="greaterOrEqual" value="60"/>
            <SimplePredicate field="fld1" operator="lessOrEqual" value="70"/>
          </CompoundPredicate>
          <Node score="tgtZ">
            <SimpleSetPredicate booleanOperator="isNotIn" field="fld4">
              <Array type="string">optA optB</Array>
            </SimpleSetPredicate>
          </Node>
        </Node>
      </Node>
      <Node score="tgtY">
        <CompoundPredicate booleanOperator="or">
          <SimplePredicate field="fld4" operator="equal" value="optA"/>
          <SimplePredicate field="fld4" operator="equal" value="optC"/>
        </CompoundPredicate>
        <Node score="tgtY">
          <CompoundPredicate booleanOperator="and">
            <SimplePredicate field="fld1" operator="greaterThan" value="10"/>
            <SimplePredicate field="fld1" operator="lessThan" value="50"/>
            <SimplePredicate field="fld4" operator="equal" value="optA"/>
            <SimplePredicate field="fld2" operator="lessThan" value="100"/>
            <SimplePredicate field="fld3" operator="equal" value="false"/>
          </CompoundPredicate>
        </Node>
        <Node score="tgtZ">
          <CompoundPredicate booleanOperator="and">
            <SimplePredicate field="fld4" operator="equal" value="optC"/>
            <SimplePredicate field="fld2" operator="lessThan" value="30"/>
          </CompoundPredicate>
        </Node>
      </Node>
    </Node>
  </TreeModel>
</PMML>
```

**Comment by vruusmann** [comment]
> but the target fld5 result of other software is "tgtY"

What is the name and version of this other software?

Your sample data record follows the path from Node@id=1 to Node@id=3, and then stops, because Node@id=4 evaluates to `false`:
```XML
<Node id="1" score="tgtX">
	<True/>
	<Node id="2" score="tgtX">
		<SimplePredicate field="fld4" operator="equal" value="optA"/>
		<Node score="tgtX">
			<CompoundPredicate booleanOperator="surrogate">
				<SimplePredicate field="fld1" operator="lessThan" value="30"/>
				<SimplePredicate field="fld2" operator="greaterThan" value="20"/>
			</CompoundPredicate>
			<Node score="tgtX">
				<SimplePredicate field="fld2" operator="lessThan" value="40"/>
			</Node>
			<Node score="tgtZ">
				<SimplePredicate field="fld2" operator="greaterOrEqual" value="10"/>
			</Node>
		</Node>
		<Node id="3" score="tgtZ">
			<CompoundPredicate booleanOperator="or">
				<SimplePredicate field="fld1" operator="greaterOrEqual" value="60"/>
				<SimplePredicate field="fld1" operator="lessOrEqual" value="70"/>
			</CompoundPredicate>
			<Node id="4" score="tgtZ">
				<SimpleSetPredicate booleanOperator="isNotIn" field="fld4">
					<Array type="string">optA optB</Array>
				</SimpleSetPredicate>
			</Node>
		</Node>
	</Node>
</Node>
```

This situation is known as "no true child", where the result is determined by the `TreeModel@noTrueChildStrategy` attribute. This attribute is not defined in your model, which means that the default value `returnNullPrediction` should be assumed.

If you want to return the "last known good"-value, then this attribute should be explicitly set to `returnLastPrediction`:
```XML
<TreeModel noTrueChildStrategy="returnLastPrediction">
</TreeModel>
```

In conclusion, the JPMML-Evaluator library is returning the correct score (ie. `null`), and this other software is returning an incorrect result.
